code/game/objects/items/devices/radio/radio_objects.dm

```diff
@@ -116,6 +117,31 @@ GLOBAL_LIST_EMPTY(deadsay_radio_systems)
 
 	for(var/ch_name in channels)
 		secure_radio_connections[ch_name] = SSradio.add_object(src, SSradio.radiochannels[ch_name],  RADIO_CHAT)
+	if(canhear_range)
+		AddComponent(/datum/component/proximity_monitor, canhear_range)
+
+/obj/item/radio/HasProximity(mob/crosser)
+	if(!istype(crosser) || (crosser in listeners))
+		return
+	listeners += crosser
+	RegisterSignal(crosser, COMSIG_MOVABLE_MOVED, PROC_REF(is_crosser_still_listening))
+	RegisterSignal(crosser, COMSIG_PARENT_QDELETING, PROC_REF(remove_from_listener_list))
+
+/obj/item/radio/proc/is_crosser_still_listening(mob/crosser)
+	SIGNAL_HANDLER
+	var/still_listening = FALSE
+	for(var/obj/effect/abstract/proximity_checker/checker in get_turf(crosser))
+		if(checker.monitor.hasprox_receiver == src)
+			still_listening = TRUE
+			break
```

Comment:
```suggestion
	for(var/obj/effect/abstract/proximity_checker/checker in get_turf(crosser))
		if(checker.monitor.hasprox_receiver == src)
			still_listening = TRUE
			break
```
This is scuffed but I don't think we have a better way to do this. It also looks way worse than it actually is